### DISCUSSION TITLE
`IonizationState` temperatures

### DISCUSSION THREAD

**Discussion by StanczakDominik** [PlasmaPy/PlasmaPy#1020] [Particles]
﻿@namurphy, here's a `IonizationState` usage question! [The docs](https://docs.plasmapy.org/en/stable/api/plasmapy.particles.ionization_state.IonizationState.html#plasmapy.particles.ionization_state.IonizationState) show that `IonizationState` has a field for the *electron* temperature. Suppose I wanted to use that to model a population of carbon impurities of various charge states. Is it reasonable to assume that `T_e` there would correspond to the (assumed in-equilibrium-with-electrons) temperature of all the carbon charge states?

I guess this is a little confusing to me because suppose we've got two sets of impurities, carbon and tungsten. You could imagine those populations being at different temperatures; but then I don't see why they'd each get an "electron" temperature. Would it not be more reasonable to assume that both carbon and tungsten should have a "species" temperature, and then you can have an electron population with a temperature that's somewhere in between as it tries to equilibrate collisionally with both of them? (probably you'd get a non-maxwellian distribution for electrons...)

All in all, seems like there's plenty of overlap between the new `Species` class from #1008, as I had imagined it (barring, maybe, particle simulations, but that's not quite what I had in mind) and the existing `IonizationState` class. In fact, I'm starting to think Species is a special case of IonizationState, with ~~`[1]`~~ EDIT: `[0, 1, 0, 0, 0, 0, 0]` for `ionic_fractions` (assuming we just want singly-ionized carbon)!

**Comment by namurphy** [comment]
Oh, good question.  The reason why I put only electron temperature in there was solely because of my use case for non-equilibrium ionization modeling in the solar wind and coronal mass ejections.  The ionization and recombination rates depend strongly on electron temperature but very weakly on ion temperature (since the ions are approximately stationary with respect to the population of electrons...especially the more massive ones).  It would be completely reasonable to have ion temperatures in there too, which will often be different than the electron temperatures.

I did create a small class, `IonicFraction` in the same file as `IonizationState` to describe one ionic level.  For example, `IonizationState(...)[0]` would give an `IonicFraction` instance for Z=0.  It seems to me that the `Species` class would be a more general replacement for `IonicFraction`.  That is to say, `IonizationState(...)[Z]` could return a `Species` instance rather than an `IonicFraction` instance.  

And for non-equilibrium ionization modeling...we've usually ignored that there could be a non-Maxwellian electrons.  @ionizationcalc has been working on calculating the relevant rates for kappa distributions.  Parker Solar Probe is informing us about the electron distribution functions in the solar wind and ejected plasma at ~ten solar radii, and so we could probably make a more informed guess about what the electron distribution functions are within ~2–3 solar radii before the ionization states freeze out. The rates for kappa distributions are going to be really helpful.

@StanczakDominik — Thank you for noticing this connection!

**Reply by StanczakDominik** [reply]
Okay then, how about this: initially T_e can be specified in IonizationState and all the `IonicFraction`s inside inherit their `T_i` from `T_e` by default (assuming thermal equilibrium); if someone wants to, they can then modify that `IonicFraction`'s T_i and get non-equilibrium situations, but it's up to them. This way, I can iterate over IonicFractions and access their temperatures and number densities without passing around their parent IonizationStates.

Obviously T_i is a bit of a stretch for [0], the neutral states, but let's live with it, I suppose.

---


Also, note how the language above doesn't really *flow*... I don't really like the name IonicFraction if it's carrying around more than a float (e.g. temperatures, densities), but I don't really like the name Species for that in the context, either. ChargeState? Probably too close conceptually to IonizationState....

Use case: For my neoclassical calculations I'm going to have to iterate over multiple isotopes (say, carbon and hydrogen for now), and for each of them I'm going to have multiple charge states. I would really like to have good vocab for this, so I can make this clean in my code.

**Reply by StanczakDominik** [reply]
... I just now realized that IonizationState doesn't **store** `IonicFraction`s, it generates them on the fly via `__getitem__`.

Well, I'll have to change that design a little for this to work, but okay :sweat_smile: To the refactor cave! :bat:

**Reply by namurphy** [reply]
> Okay then, how about this: initially T_e can be specified in IonizationState and all the `IonicFraction`s inside inherit their `T_i` from `T_e` by default

This seems like a reasonable assumption to me, so long as we make sure this assumption is well-documented.

> Obviously T_i is a bit of a stretch for [0], the neutral states, but let's live with it, I suppose.

I am neutral on this issue.

> I don't really like the name IonicFraction

Yeah, I haven't been happy with this either.  Before my last refactoring, it was `State` (I think) which was too vague.  Thinking about this more...maybe `IonicLevel`?  That seems at about the right balance between ~~vagueness~~ abstraction and specificity.

> Use case: For my neoclassical calculations I'm going to have to iterate over multiple isotopes (say, carbon and hydrogen for now), and for each of them I'm going to have multiple charge states. 

That seems like a great use case for `IonizationStateCollection`!

> I would really like to have good vocab for this, so I can make this clean in my code.

Yeah...I've noticed that the vocabulary around this differs somewhat between solar physics and fusion energy sciences, so we're going to have to think carefully about the language we can use that is common to both fields.  Either that, or we should decide what language _we think_ should be common to both fields!

**Comment by StanczakDominik** [comment]
Another question:

https://github.com/PlasmaPy/PlasmaPy/blob/af7b9d3b62488a5bf99f7d054a9e6bdfe17f50fd/plasmapy/particles/ionization_state.py#L231

Is this as "simple" as:

1. assuming N0 = 1
2. plugging in [the Saha equation](https://docs.plasmapy.org/en/latest/api/plasmapy.formulary.ionization.Saha.html) to calculate Ni for i in charge levels
3. renormalizing N to sum to the full number density?

Looks like the only thing we'd really miss here would be the degeneracies between the ionization states. Is there a systematic way of deriving them? If not, should we attach them to our big particle json?
